luarules/gadgets/cmd_build_bugger_off.lua

```diff
@@ -1,186 +1,193 @@
-                        end    
-                    end
-                end
-            end
-        end
-        if frame % 100 ~= 0 then
-            return
-        end
-        for builderID, _ in pairs(slowUpdateBuilders) do
-            -- local firstCommand = getFirstCommand(builderID)
-            local cmdID, options, tag, targetX, targetY, targetZ = Spring.GetUnitCurrentCommand(builderID, 1)
-            local x, y, z = Spring.GetUnitPosition(builderID)
-            if cmdID == nil or cmdID > -1 or isBuilding then
-                print("Clearing watched builder slow")
-                slowUpdateBuilders[builderID] = nil
-                builderRadiusOffsets[builderID] = nil
-            elseif distance(targetX, targetZ, x, z) <= FAST_UPDATE_RADIUS then
-                print("Promote to fast")
-                slowUpdateBuilders[builderID] = nil
-                watchedBuilders[builderID] = true
-            end
-
-        end
-
-    end
-
-    function gadget:AllowCommand(unitID, unitDefID, teamID, cmdID, cmdParams, cmdOptions, cmdTag, playerID, fromSynced, fromLua)
-        local orderedUnit = unitID
-
-        local unitDefID = Spring.GetUnitDefID(orderedUnit)
-        if cachedUnitDefs[unitDefID].isBuilder then
-            watchedBuilders[orderedUnit] = true
-            builderRadiusOffsets[orderedUnit] = 1
-        end
-        return true
-
-    end
+	local shouldNotBuggeroff = {}
+	local cachedUnitDefs = {}
+	local cachedBuilderTeams = {}
+	for unitDefID, unitDef in pairs(UnitDefs) do
+		if unitDef.isImmobile then
+			shouldNotBuggeroff[unitDefID] = true
+		end
+		
+		cachedUnitDefs[unitDefID] = { radius = unitDef.radius, isBuilder = unitDef.isBuilder}
+	end
+
+	local function willBeNearTarget(unitID, tx, ty, tz, seconds, maxDistance)
+		local ux, uy, uz = Spring.GetUnitPosition(unitID)
+		if not ux then return false end
+		
+		local vx, vy, vz = Spring.GetUnitVelocity(unitID)
+		if not vx then return false end
+		
+		local futureX = ux + vx * seconds * Game.gameSpeed
+		local futureY = uy + vy * seconds * Game.gameSpeed
+		local futureZ = uz + vz * seconds * Game.gameSpeed
+		
+		local dx = futureX - tx
+		local dy = futureY - ty
+		local dz = futureZ - tz
+		return math.diag(dx, dy, dz) <= maxDistance
+	end
+
+	local function isInTargetArea(interferingUnitID, x, y, z, radius)
+		local ux, uy, uz = Spring.GetUnitPosition(interferingUnitID)
+		if not ux then return false end
+		return math.diag(ux - x, uz - z) <= radius
+	end
+	
+	local slowUpdateBuilders = {}
+	local watchedBuilders = {}
+	local builderRadiusOffsets = {}
+
+	local FAST_UPDATE_RADIUS	= 400
+	-- builders take about this much to enter build stance; determined empirically
+	local BUILDER_DELAY_SECONDS = 3.3
+	local BUILDER_BUILD_RADIUS  = 200
+	local SEARCH_RADIUS_OFFSET  = 200
+	local FAST_UPDATE_FREQUENCY = 30
+	local SLOW_UPDATE_FREQUENCY = 100
+	local BUGGEROFF_RADIUS_ESCALATION_INCREMENT = FAST_UPDATE_FREQUENCY * 0.5
```

Comment:
```suggestion
	local SEARCH_RADIUS_INCREMENT = FAST_UPDATE_FREQUENCY * 0.5
```
Shorter and matches naming of `SEARCH_RADIUS_OFFSET`

Reply:
Hmm, but that's not quite what it is. The search radius and the area to be cleared are different.
You look in a circle of size + SEARCH_RADIUS_OFFSET to enumerate all the interfering units, and tell them to get out of a circle of size + BUGGEROFF_RADIUS_ESCALATION_INCREMENT * the number of tries.
Maybe CLEAR_RADIUS_INCREMENT?

Reply:
`BUGGEROFF_RADIUS_INCREMENT ` is fine